src/dvclive/monitor_system.py

```diff
@@ -0,0 +1,158 @@
+import logging
+from typing import Dict, Union, Optional, List, Tuple
+from pathlib import Path
+
+import psutil
+from statistics import mean
+from threading import Event, Thread
+from funcy import merge_with
+
+from .error import InvalidDataTypeError
+
+
+logger = logging.getLogger("dvclive")
+MEGABYTES_DIVIDER = 1024.0**2
+GIGABYTES_DIVIDER = 1024.0**3
+
+MINIMUM_CPU_USAGE_TO_BE_ACTIVE = 20
+
+
+class _MonitorSystem:
+    """
+    Monitor system resources and log them to DVC Live.
+    Use a separate thread to call a `_get_metrics` function at  fix interval and
+    aggregate the results of this sampling using the average.
+    """
+
+    _plot_blacklist_prefix: Tuple = ()
+
+    def __init__(
+        self,
+        interval: float = 0.5,
+        num_samples: int = 10,
+        plot: bool = True,
+    ):
+        if not isinstance(interval, (int, float)):
+            raise InvalidDataTypeError("MonitorSystem.interval", type(interval))
+        if not isinstance(num_samples, int):
+            raise InvalidDataTypeError("MonitorSystem.sample", type(num_samples))
+        if not isinstance(plot, bool):
+            raise InvalidDataTypeError("MonitorSystem.plot", type(plot))
+
+        self._interval = interval  # seconds
+        self._nb_samples = num_samples
+        self._plot = plot
+        self._warn_user = True
+
+    def __call__(self, live):
+        self._live = live
+        self._shutdown_event = Event()
+        Thread(
+            target=self._monitoring_loop,
+        ).start()
+
+    def _monitoring_loop(self):
+        while not self._shutdown_event.is_set():
+            self._metrics = {}
+            for _ in range(self._nb_samples):
+                last_metrics = {}
+                try:
+                    last_metrics = self._get_metrics()
+                except psutil.Error:
+                    if self._warn_user:
+                        logger.exception("Failed to monitor CPU metrics")
+                        self._warn_user = False
+
+                self._metrics = merge_with(sum, self._metrics, last_metrics)
+                self._shutdown_event.wait(self._interval)
+                if self._shutdown_event.is_set():
+                    break
+            for name, values in self._metrics.items():
+                blacklisted = any(
+                    name.startswith(prefix) for prefix in self._plot_blacklist_prefix
```

Comment:
I think it's better to have exact metrics names vs startswith. startswith can lead to unexpected false positives later. Was there a strong reason to use it?

Reply:
I agree with the general conclusion on the false positives. But as soon as we allow the user to specify paths in the metric name, we are forced to test only the beginning of the name.
For instance, the total amount of memory for disk usage looks like this: `system/disks/total (GB)/<path from the user>`. I cannot check the full metric name in that case. 

What I could do is check the full path for metrics names that are not user dependent. Then add the `startswith` check for the metrics name that are user dependant (like disk usage)

Reply:
kk, I think it should be fine in this case since we control all these names.